src/coq/decompiler/decompiler.ml

```diff
@@ -63,14 +92,25 @@ let run_tac env sigma (tac : unit Proofview.tactic) (goal : constr)
     
 (* Returns true if the given tactic solves the goal. *)
 let solves env sigma (tac : unit Proofview.tactic) (goal : constr) =
-  fst (run_tac env sigma tac goal) = []
+  let subgoals, _ = run_tac env sigma tac goal in subgoals = []
 
 (* Compute the type of a term if possible, otherwise None. *)
 let type_of env (trm : constr) : types option =
   try Some (Typeops.infer env trm).uj_type
   with _ -> None
-  
-            
+
+let subterm_with_type env sigma : types -> constr -> (Evd.evar_map * (env * constr)) list =
+  Hofs.map_term_env_if_list
+    (fun env sigma typ trm ->
+      sigma, match type_of env trm with
+      | Some t -> equal t typ
```

Comment:
Hm, so this will require exact syntactic equality with the type. Is that what you want? Or do you want proper type checking? If the former, why is that sufficient? If the latter, you probably want `has_type` from checking.ml.

Reply:
Good point, fixed